src/coreclr/jit/codegencommon.cpp

```diff
@@ -390,15 +390,33 @@ void CodeGen::genMarkLabelsForCodegen()
                 break;
 
             case BBJ_COND:
-                JITDUMP("  " FMT_BB " : branch target\n", block->GetTrueTarget()->bbNum);
-                block->GetTrueTarget()->SetFlags(BBF_HAS_LABEL);
-
-                // If we need a jump to the false target, give it a label
-                if (!block->CanRemoveJumpToTarget(block->GetFalseTarget(), compiler))
+                // Check if reversing the condition will allow us to fall into the false target
+                if (block->CanRemoveJumpToTarget(block->GetTrueTarget(), compiler))
+                {
+                    compiler->gtReverseCond(block->lastNode());
```

Comment:
gtReverseCond is not safe for NaNs

Reply:
Does that mean we need something like `gtReverseCond(test->AsOp()->gtOp1)`? That's the pattern we use e.g. during block reordering.

Reply:
Ah, that pattern triggers asserts -- I'm guessing it would make sense to do this opt before lowering, like maybe in `fgUpdateFlowGraph` when optimizing layout, though ideally we'd do this when layout is unlikely to change.

Reply:
I mean that you can just guard it with `!varTypeIsFloating`, Although, looks like `gtReverseCond` does something for floats inside, just not sure if it's correct

Reply:
Reversing float compares may end up with something that requires generating multiple jumps, so this may also deoptimize those, although not sure if it's balanced out by what is being optimized here. You may need to swap the operands after, take a look at `GenCondition::PreferSwap`.

However, to me this seems like an odd place to be making flow graph changes given the name of the function and its existing logic. If we do it here the name/documentation of the function should be updated.

Reply:
> gtReverseCond is not safe for NaNs

`gtReverseCond` should work just fine for FP compares, since it switches whether the comparison is ordered or not.

Reply:
> Reversing float compares may end up with something that requires generating multiple jumps, so this may also deoptimize those, although not sure if it's balanced out by what is being optimized here.

According to the [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=617790&view=ms.vss-build-web.run-extensions-tab), there are some regressions on x86/x64, though these size regressions are due to larger jumps, but fewer jumps overall.

> You may need to swap the operands after, take a look at GenCondition::PreferSwap.

Is it possible to create a `GenCondition` to do this check after lowering? I'm hitting asserts like `OperIsSimple`, `OperIsCompare`, etc, so I'm guessing this only works on frontend nodes.

> However, to me this seems like an odd place to be making flow graph changes given the name of the function and its existing logic. If we do it here the name/documentation of the function should be updated.

I agree it's an odd place. I wasn't getting ideal results doing this in `fgUpdateFlowGraph`, as the flowgraph transformations it does sometimes limit/invalidate prior decisions to reverse conditions. I could add a new phase after `optOptimizeLayout` to do these reversals, but I'd prefer not to introduce another iteration over the blocklist just for this. We already have some checks for jump removal in `genMarkLabelsForCodegen` so we don't create unnecessary labels; I can document this logic better. Any suggestions for a new name?

> Also, consider asserting that the reversed tree is the same as the original tree.

Sure thing.

Reply:
> Is it possible to create a `GenCondition` to do this check after lowering? I'm hitting asserts like `OperIsSimple`, `OperIsCompare`, etc, so I'm guessing this only works on frontend nodes.

You would have to find the compare node and swap its operands. It does seem non-trivial at this point.

> 
> > However, to me this seems like an odd place to be making flow graph changes given the name of the function and its existing logic. If we do it here the name/documentation of the function should be updated.
> 
> I agree it's an odd place. I wasn't getting ideal results doing this in `fgUpdateFlowGraph`, as the flowgraph transformations it does sometimes limit/invalidate prior decisions to reverse conditions. I could add a new phase after `optOptimizeLayout` to do these reversals, but I'd prefer not to introduce another iteration over the blocklist just for this. We already have some checks for jump removal in `genMarkLabelsForCodegen` so we don't create unnecessary labels; I can document this logic better. Any suggestions for a new name?

Is there any significant cost in doing this as a separate traversal (limited to FullOpts)? If not I would much prefer that to keep things orthogonal. `genMarkLabelsForCodegen` is nicely self contained with one responsibility today.

The fact that this works better at this point seems to be an indication that other flow graph related optimizations would make sense to do at this point in the JIT pipeline. If I'm understanding correctly, this point is actually right after LSRA and alignment but before any codegen. Is that also the point where the new block layout algorithm is going to be? If so, that seems like motivation to just introduce the phase now and get started on the post-layout fixups. I assume the block reordering will uncover many cases like these where we should reverse branches, but it would make more sense to me to do these as part of its phase.

Reply:
> Is there any significant cost in doing this as a separate traversal (limited to FullOpts)? If not I would much prefer that to keep things orthogonal.

Let me give this a try.

> The fact that this works better at this point seems to be an indication that other flow graph related optimizations would make sense to do at this point in the JIT pipeline. If I'm understanding correctly, this point is actually right after LSRA and alignment but before any codegen.

That's correct.

> Is that also the point where the new block layout algorithm is going to be?

I'm not sure yet, though that is a consideration in #93020. LSRA can create new blocks, though I don't think it happens that often, and we might be able to tweak LSRA to place these blocks in sensible locations. We can also remove blocks when setting stack levels, which may create the opportunity for `BBJ_COND` fallthrough behavior if we reverse the condition (even if we don't do block reordering after LSRA, maybe we can at least do it after stack level setting?).